src/MMCPClient.cpp

```diff
@@ -0,0 +1,763 @@
+    }
+
+    if (mpHost->getMMCPAllowConnectionRequests()) {
+        const QString infoMsg = tr("[ CHAT ]  - %1 has requested your public connections...").arg(m_chatName);
+        mpHost->postMessage(infoMsg);
+
+        server->sendPublicConnections(this);
+    } else {
+        const QString infoMsg = tr("[ CHAT ]  - %1 has requested your public connections, but you're ignoring connection requests...").arg(m_chatName);
+        mpHost->postMessage(infoMsg);
+    }
+}
+
+/**
+ * Display a chat all message and echo it to any clients we may be serving
+ */
+void MMCPClient::handleIncomingChatEveryone(const QString& msg)
+{
+    if (m_isIgnored)
+        return;
+
+    server->clientMessage(msg);
+
+    if (m_isServed) {
+        //Echo this message to
+        server->sendServedMessage(this, msg);
+    } else {
+        //Echo message to other clients we might be serving
+        server->sendMessageToServed(this, msg);
+    }
+}
+
+/**
+ * Display an incoming personal chat
+ */
+void MMCPClient::handleIncomingChatPersonal(const QString& msg)
+{
+    server->clientMessage(msg);
+}
+
+/**
+ * Display an incoming chat to a group
+ */
+void MMCPClient::handleIncomingChatGroup(const QString& msg)
+{
+    const QString groupStr = msg.left(15).trimmed();
+    const QString trimmedMsg = msg.right(msg.length() - 15);
+
+    using namespace AnsiColors;
+
+    const QString groupMsg = QString("%1%2(%3%4%5)%6")
+                                .arg(RST).arg(FBLDRED).arg(FBLDCYN)
+                                .arg(groupStr).arg(FBLDRED)
+                                .arg(trimmedMsg);
+
+    server->clientMessage(groupMsg);
+}
+
+/**
+ * A client changed their name
+ */
+void MMCPClient::handleIncomingNameChange(const QString& newName)
+{
+    const QString infoMsg = tr("[ CHAT ]  - %1 is now known as %2.").arg(m_chatName).arg(newName);
+    mpHost->postMessage(infoMsg);
+    m_chatName = newName;
+}
+
+/**
+ * Someone has requested to peek our connections
+ */
+void MMCPClient::handleIncomingPeekConnections()
+{
+    //check if this client is ignored before doing anything drastic
+    if (m_isIgnored) {
+        const QString infoMsg = tr("[ CHAT ]  - %1 is trying to peek your connections!").arg(m_chatName);
+        mpHost->postMessage(infoMsg);
+        return;
+    }
+
+    if (mpHost->getMMCPAllowPeekRequests()) {
+        const QString infoMsg = tr("[ CHAT ]  - %1 is peeking at your connections...").arg(m_chatName);
+        mpHost->postMessage(infoMsg);
+
+        server->sendPublicPeek(this);
+    } else {
+        const QString infoMsg = tr("[ CHAT ]  - %1 is trying to peek your connections, but you're ignoring peek requests...").arg(m_chatName);
+        mpHost->postMessage(infoMsg);
+    }
+}
+
+/**
+ * Display someones peek list that we've requested
+ */
+void MMCPClient::handleIncomingPeekList(const QString& list)
+{
+    server->clientMessage(list);
+
+    QStringList parts = list.split("~");
+
+    if (parts.size() % 3 != 0) {
```

Comment:
This doesn't work correctly - for a single connection I had a string `127.0.0.1~1051~Mudlet~` because of that third `~` necessary to terminate THAT entry from a following one (which there isn't) in this case then the string gets split into 4 (four) parts - but because there is nothing following it the forth part is an empty string. However that does not give a zero result when modded with `3`...

I'm also trying to work out why I'm getting a `1051`port number from a TinTin++ peer but I think that is a problem/bug in *that* client.

Reply:
Ah yep thanks for pointing that out